src/steps/navigationSteps.ts

```diff
@@ -0,0 +1,121 @@
+import { config } from './../config';
+import {getNavigationState, getStepsFromNavigation, steps } from './steps';
+
+
+// We only report the active steps in the last 2 navigation step to prevent
+// Convert the steps.navigationSteps to an array of active steps.
+// "stale steps" from being reported.
+//
+// "stale steps" are defined as steps where the start mark was executed but the
+// corresponding end mark was never executed - thus leaving it as a "stale step"
+export const getActiveStepsFromNavigationSteps = (): Record<string, boolean>  => {
+  const navigationSteps = getNavigationState();
+  const { startMarkToStepsMap } = steps;
+  if (Object.keys(navigationSteps).length === 0) {
+    return {};
+  }
+
+  const activeSteps: Record<string, boolean> = {};
+  const currentNavIndex = Object.keys(navigationSteps).length - 1;
+
+  const currentNavStep = getStepsFromNavigation(currentNavIndex);
+
+  Object.keys(currentNavStep).forEach((startMark: string) => {
+    const ongoingSteps = startMarkToStepsMap[startMark] ?? [];
+    Object.keys(ongoingSteps).forEach((step) => {
+      activeSteps[step] = true;
+    });
+  });
+
+  // if nav length is >1, then check the nav step prior as well
+  if (Object.keys(navigationSteps).length > 1) {
+    const prevNavStep = getStepsFromNavigation(currentNavIndex - 1);
+
+    Object.keys(prevNavStep).forEach((startMark: string) => {
+      const prevNavSteps = startMarkToStepsMap[startMark] ?? [];
+      Object.keys(prevNavSteps).forEach((step) => {
+        activeSteps[step] = true;
+      });
+    });
+  }
+
+  return activeSteps;
+}
+
+// start mark for navigationSteps data structure
+export const recordStartMark = (startMark: string) => {
+  const { navigationSteps } = steps;
+
+  // if we don't have any initial states recorded, initialize one
+  const navigationLength = Object.keys(navigationSteps).length > 0 ? Object.keys(navigationSteps).length : 1;
```

Comment:
If you need this `Object.keys(navigationSteps).length` anyway, just store it in a variable, we save some extra byte.